File: Worker/Worker.py
import threading
import multiprocessing
from datetime import datetime
from common import Config, printheader, printsuccess, printdebug, printerror, getTime, excpetion_info
from confluent_kafka import Consumer, Producer, TopicPartition

class Task(multiprocessing.Process):
    task_id = 0

    def __init__(self, task, worker_success_value):
        multiprocessing.Process.__init__(self)
        self.task = task
        self.worker_success_value = worker_success_value
        self.task_id = Task.task_id
        Task.task_id += 1
    
    def run(self):
        def __get_header(headers_tuple, key):
            for header in headers_tuple:
                if header[0] == key:
                    return header[1]
            return None

        def __set_header(headers_tuples, new_tuples):
            new_keys = [i[0] for i in new_tuples]
            return new_tuples + [i for i in headers_tuples if i[0] not in new_keys]

        def __get_topic(headers_tuple):
            try:
                headers = dict((k, v.decode('utf-8')) for k, v in headers_tuple)
                if (not headers) or (not headers['topic']):
                    return False
                if (not headers['time']):
                    return headers['topic']

                time_diff = (datetime.strptime(headers['time'], Config.get('sm_time_format')) - datetime.now()).total_seconds()
                
                if time_diff < -1 * Config.get('sm_miniumum_delay'):
                    printerror(f'Task.run(): WORKER DELAY of {time_diff}: {headers}')

                # If the message is scheduled for a time too small to bucket, we send it to the provided topic
                if time_diff < Config.get('sm_miniumum_delay'):
                    return headers['topic']
                
                # If the message is scheduled for the future (+ bucket process time), then we need to bucket it in the associated bucket
                for bucket in reversed(Config.get('bucket_object_list')):
                    if time_diff > (1 + Config.get('sm_bucket_process_time_fraction')) * bucket['lower']:
                        return bucket['name']

                return headers['topic']
            except Exception as e:
                printerror(f'Unable to get topic name from headers: {headers}')
                printerror(e)
                return False

        def __run():
            
            group_id = Config.get('sm_consumer_group_prefix') + self.task['job']['name'] + '_' + str(self.task_id)
            
            consumer = Consumer({
                'bootstrap.servers':    Config.get('kafka_server'),
                'group.id':             group_id,
                'enable.auto.commit':   True,
                'auto.offset.reset':    'earliest',
            })

            producer = Producer({
                'bootstrap.servers':    Config.get('kafka_server')})
            
            topic_partition = TopicPartition(self.task['job']['name'], self.task_id)
            consumer.assign([topic_partition])

            while True:

                message = consumer.poll(timeout=Config.get('worker_consumer_timeout'))
                
                if message is None or message.error():
                    message and message.error() and printerror(f'Task.run() MessageError: {message.error()}')
                    break
                
                headers = message.headers()
                topic = __get_topic(headers)
                hop_count = __get_header(headers, Config.get('sm_header_message_hopcount_key'))
                hop_count = (int(hop_count) + 1) if hop_count else 0

                if topic:
                    message_job_id = __get_header(headers, Config.get('sm_header_job_id_key'))

                    message.set_headers(__set_header(message.headers(), [(Config.get('sm_header_job_id_key'), bytes(self.task['job_id'], 'utf-8')),
                                         (Config.get('sm_header_worker_timestamp_key'), bytes(getTime(), 'utf-8')),
                                         (Config.get('sm_header_message_hopcount_key'), bytes(str(hop_count), 'utf-8'))]))

                    producer.produce(
                        topic   = topic,
                        value   = message.value(),
                        key     = message.key(),
                        headers = message.headers(),
                    )
                    
                    if bytes(self.task['job_id'], 'utf-8') == message_job_id:
                        break

            consumer.close()
            producer.flush()


        try:
            __run()
        except Exception as e:
            printerror(f'Task Failed in thread ID: {threading.get_native_id()}')
            printerror(e)
            self.worker_success_value.value = False

class Worker(multiprocessing.Process):
    def __init__(self, work, WorkerHandler):
        multiprocessing.Process.__init__(self)
        self.task = work
        self.worker_handler = WorkerHandler
        self.worker_success_value = multiprocessing.Value('i', 1)

    def run(self):
        TaskProcessList = []
        # One process per partition
        for _ in range(Config.get('sm_partitions_per_bucket')):
            t = Task(self.task, self.worker_success_value)
            t.start()
            TaskProcessList.append(t)
        
        for t in TaskProcessList:
            t.join()
        
        self.task['status'] = Config.get('worker_status_list.done')
        
        if not self.worker_success_value.value:
            self.task['status'] = Config.get('worker_status_list.error')
            printerror(f'Worker.run(): Worker failed {self.task}')
        
        self.worker_handler.post_worker(self.task)

if __name__=='__main__':
    printheader('Worker Service')
    

File: Worker/src/worker.rs
use rdkafka::config::ClientConfig;
use rdkafka::consumer::base_consumer::BaseConsumer;
use rdkafka::message::{Headers, BorrowedHeaders};
use rdkafka::util::Timeout;
use rdkafka::producer::{BaseProducer, BaseRecord};
use rdkafka::error::{KafkaError, RDKafkaErrorCode};

use rdkafka::consumer::Consumer;
use rdkafka::message::Message;
use rdkafka::topic_partition_list::TopicPartitionList;

use serde_json::Value as Json;
use serde_json::json;

use std::process::exit;
use std::time::{Duration, Instant};

use crate::constants::Constants;
use crate::common::*;
use crate::multiprocess as mp;

struct Task {
    task : Json,
    config : Json,
    worker_success : mp::SharedMemory<bool>,
    task_id : i32,
}

impl Clone for Task {
    fn clone(&self) -> Task {
        Task {
            task: self.task.clone(),
            config: self.config.clone(),
            worker_success: self.worker_success.clone(),
            task_id: self.task_id,
        }
    }
}

impl Task {
    pub fn new(task : Json, config : Json, worker_success : mp::SharedMemory<bool>, task_id : i32) -> Task {
        Task {
            task,
            config,
            worker_success,
            task_id
        }
    }
    
    pub fn run(&self) -> mp::JoinHandle {
        let __run = | task_obj : Task | {
            let __task_run = | task_obj : Task | -> Result<(), String> {
                let __get_topic_name = |config: &Json, headers: &BorrowedHeaders| -> Option<String> {
                    let mut sm_topic_name = None;
                    let mut sm_time = None;
            
                    for i in 0..headers.count() {
                        let header = headers.get(i).unwrap_or_default();
                        if header.0 == "topic" {
                            sm_topic_name = Some(u8_to_str(header.1));
                        } else if header.0 == "time" {
                            sm_time = Some(u8_to_str(header.1));
                        }
                    }
                    if sm_topic_name == None {
                        return None;
                    }
                    if sm_time == None {
                        return sm_topic_name;
                    }
            
                    let sm_time = sm_time.unwrap();
                    let sm_time = chrono::NaiveDateTime::parse_from_str(sm_time.as_str(), get_string(&config, "sm_time_format").as_str()).unwrap();
                    let time_diff = (sm_time - chrono::Local::now().naive_local()).num_seconds();
                    
                    if time_diff < -1 * get_number(&config, "sm_miniumum_delay") as i64 {
                        let error_msg = format!("WORKER DELAY: {} secs : {:?}", time_diff, sm_time);
                        print_error(error_msg.as_str());
                        return sm_topic_name;
                    }
                    if time_diff < get_number(&config, "sm_miniumum_delay") as i64 {
                        return sm_topic_name;
                    }
                    let bucket_list = get_list(&config, "bucket_object_list").into_iter().rev().collect::<Vec<_>>();
                    for bucket in bucket_list.iter() {
                        let b_time = (1.0 + get_float(&config, "sm_bucket_process_time_fraction")) * get_float(&bucket, "lower");
                        if time_diff > b_time as i64 {
                            return Some(get_string(bucket, "name"));
                        }
                    }
                    sm_topic_name
                };

                let __get_header = |headers: &BorrowedHeaders, key: &str| -> Option<String> {
                    for i in 0..headers.count() {
                        let header = headers.get(i).unwrap_or_default();
                        if header.0 == key {
                            return Some(u8_to_str(header.1));
                        }
                    }
                    None
                };

                let __get_hop_count = |headers: &BorrowedHeaders, config: &Json| -> i32 {
                    let key = get_string(&config, "sm_header_message_hopcount_key");
                    let hop_count = __get_header(headers, key.as_str());
                    if hop_count == None {
                        return 0;
                    }
                    hop_count.unwrap().parse::<i32>().unwrap()
                };

                let __get_job_id = |headers: &BorrowedHeaders, config: &Json| -> String {
                    let key = get_string(&config, "sm_header_job_id_key");
                    let job_id = __get_header(headers, key.as_str());
                    if job_id == None {
                        return "".to_string();
                    }
                    job_id.unwrap()
                };

                let task = task_obj.task.clone();
                let task_id = task_obj.task_id;
                let config = task_obj.config;
                let job_ob = get_object(&task, "job");
                let job_name = job_ob["name"].as_str().unwrap();
                let group_id = format!("{}{}_{}", get_string(&config, "sm_consumer_group_prefix"), job_name, task_id);

                let consumer : BaseConsumer = ClientConfig::new()
                    .set("group.id", group_id)
                    .set("bootstrap.servers", get_string(&config, "kafka_server"))
                    .set("enable.auto.commit", "true")
                    .set("auto.offset.reset", "earliest").create().expect("Task.__run(): Consumer creation error");

                let producer : BaseProducer = ClientConfig::new()
                    .set("bootstrap.servers", get_string(&config, "kafka_server"))
                    .create().expect("Task.__run(): Producer creation error");

                let mut tp_list = TopicPartitionList::new();
                tp_list.add_partition(job_name, task_id);
                consumer.assign(&tp_list).expect("Task.__run(): Topic Assignment error");

                loop {
                    match consumer.poll(std::time::Duration::from_secs(get_number(&config, "worker_consumer_timeout"))) {
                        Some(Ok(message)) => {
                            let headers = message.headers().unwrap();

                            let topic_name = __get_topic_name(&config, &headers);
                            if topic_name == None {
                                continue;
                            }
                            let topic_name = topic_name.unwrap();
                            let msg_job_id = __get_job_id(&headers, &config);
                            let msg_hop_count = (__get_hop_count(&headers, &config) + 1) as i32;

                            let msg_headers = headers.detach().add(&get_string(&config, "sm_header_job_id_key"), get_string(&task, "job_id").as_bytes())
                                .add(&get_string(&config, "sm_header_message_hopcount_key"), format!("{}", msg_hop_count).as_bytes())
                                .add(&get_string(&config, "sm_header_worker_timestamp_key"), get_time().as_bytes());

                            let msg_payload = message.payload().unwrap_or_default();
                            let msg_key = message.key().unwrap_or_default();

                            match {
                                let mut retry_count = 0;
                                
                                loop {
                                    let base_record = BaseRecord::to(topic_name.as_str()).payload(msg_payload).key(msg_key).headers(msg_headers.clone()).partition(task_id);

                                    match producer.send(base_record) {
                                        Ok(_) => break Ok(()),
                                        Err((e, record)) => {
                                            retry_count += 1;
                                            if retry_count >= 10000 {
                                                break Err((e, record));
                                            }
                                            std::thread::sleep(std::time::Duration::from_millis(100));
                                        }
                                    }
                                }
                            } {
                                Ok(_) => {
                                    if msg_job_id == get_string(&task, "job_id") {
                                        break;
                                    }
                                },
                                Err(e) => {
                                    let err_msg = format!("Task.__run(): Producer send error: {:?}", e);
                                    print_error(&err_msg);
                                    task_obj.worker_success.set(false);
                                    return Err(err_msg);
                                }
                            }
                        },
                        Some(Err(_)) => {
                            continue;
                        },
                        None => {
                            break;
                        }
                    }
                }

                Ok(())
            };
            let worker_success = task_obj.worker_success.clone();
            match __task_run(task_obj) {
                Ok(_) => {},
                Err(e) => {
                    let error_msg = format!("Task.__run: {}", e);
                    print_error(error_msg.as_str());
                    worker_success.set(false);
                }
            }
            exit(0);
        };
        mp::process(__run, self.clone())
    }
}

pub struct Worker {
    config : Json,
    constants : Constants,
    worker_success : mp::SharedMemory<bool>,
    work : Json
}

impl Clone for Worker {
    fn clone(&self) -> Worker {
        Worker {
            config: self.config.clone(),
            constants: self.constants.clone(),
            worker_success: self.worker_success.clone(),
            work: self.work.clone()
        }
    }
}

impl Worker {
    pub fn new(config : Json, constants : Constants, work : Json) -> Worker {
        match mp::SharedMemory::new(true) {
            Ok(worker_success) => {
                Worker {
                    config,
                    constants,
                    worker_success,
                    work
                }
            }
            Err(e) => {
                print_error(&format!("{}", e));
                panic!("Unable to create worker_success shared memory");
            }
        }
    }
    pub fn run(&self) -> mp::JoinHandle {
        let __run = |worker_obj : Worker| {
            let mut tasks = Vec::new();
            for task_id in 0..worker_obj.config["sm_partitions_per_bucket"].as_i64().unwrap_or(16) {
                let task = Task::new(worker_obj.work.clone(), worker_obj.config.clone(), worker_obj.worker_success.clone(), task_id as i32).run();
                tasks.push(task);
            }
            for task in tasks {
                task.join();
            }
            let mut worker_obj = worker_obj.clone();
            worker_obj.work["status"] = {
                if worker_obj.worker_success.get() {
                    json!(worker_obj.config["worker_status_list.done"].as_str().unwrap_or("DONE"))
                } else {
                    print_error(&format!("Worker failed"));
                    json!(worker_obj.config["worker_status_list.error"].as_str().unwrap_or("ERROR"))
                }
            };
            post_worker(&worker_obj.config, &worker_obj.constants, &worker_obj.work);
            exit(0);
        };
        mp::process(__run, self.clone())
    }
}